### PR TITLE
added cypress

### PR DESCRIPTION
**Pull request by Chenxi96**
added function to bypass authjs

added a function to return mock auth object also implement cypress test

edit package json and p/id

added cypress in github action

added build and run for cypress

added env NEXT_PUBLIC_USE_TEST

Update tests.yml

added prism url steps env

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [x] I have added thorough tests. (If tests are not applicable, explain why in the next section.)
- [x] Is there only 1 commit.

## Describe your changes
1. Added cypress E2E testing to this project
2. Created new file to bypass authjs for testing purposes.
3. Edit Github action workflow to have cypress work in the workflow

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -33,3 +33,13 @@ jobs:
       - run: npm run test:ci
         env:
           POSTGRES_PRISMA_URL: postgresql://postgres:password@localhost:5432/court-share?schema=public
+      - uses: cypress-io/github-action@v6
```
Comment (Chenxi96) [inline .github/workflows/tests.yml:36]: Added another step on the workflow to test E2E

File: lib/authSession.ts
Comment (Chenxi96) [inline lib/authSession.ts:1]: Created a file to return user object if `NEXT_PUBLIC_USE_TEST` is true or return `auth(req or context)`

File: cypress/e2e/HomePage.cy.ts
Comment (Chenxi96) [inline cypress/e2e/HomePage.cy.ts:1]: This test will go to the main page an click  on new post then it will fill out the form. when the form is filled it will send this to the back end to be created in prisma. It also test if the created post to see if it shows on the main page

File: cypress/e2e/HomePage.cy.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    cy.visit('/')
+    cy.get('.createPost').click()
+    cy.get('#title').type('ball runs')
+    cy.get('#location').type('Air Canada Centre, 50 Bay St. #500, Toronto, ON M5J 2L2')
+    cy.get('#description').type('We need 4 more')
+    cy.get('#availableSpots').type('4')
+    cy.get('[type="datetime-local"]').type('2024-10-01T08:30')
+    cy.get('[type="submit"]').click()
+    cy.wait(500)
+    cy.get('.jsx-7114bf14ed1cdbc > p').contains('Post created')
```
Comment (alvinc21) [inline cypress/e2e/HomePage.cy.ts]: This looks like it could fail in the future because you are likely searching for a generated frontend class name that changes.

File: cypress/support/commands.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+// declare global {
+//   namespace Cypress {
+//     interface Chainable {
+//       login(email: string, password: string): Chainable<void>
+//       drag(subject: string, options?: Partial<TypeOptions>): Chainable<Element>
+//       dismiss(subject: string, options?: Partial<TypeOptions>): Chainable<Element>
+//       visit(originalFn: CommandOriginalFn, url: string, options: Partial<VisitOptions>): Chainable<Element>
+//     }
+//   }
+// }
```
Comment (alvinc21) [inline cypress/support/commands.ts]: Is this file needed?

File: cypress/support/e2e.ts [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+//
+// You can read more here:
+// https://on.cypress.io/configuration
+// ***********************************************************
+
+// Import commands.js using ES2015 syntax:
+import './commands'
+
+// Alternatively you can use CommonJS syntax:
+// require('./commands')
```
Comment (alvinc21) [inline cypress/support/e2e.ts]: Is this file needed?

File: pages/index.tsx [hunk 5]
```diff
@@ -3,11 +3,33 @@ import { GetServerSideProps } from "next"
...
+                data: {
+                    name: 'admin',
+                    email: 'testing@example.com',
+                    image: null
+                }
+            })
+        }
+      }
+    checkAndCreateUser()
+  }
```
Comment (alvinc21) [inline pages/index.tsx]: Test assets like this user should be created and destroyed within [the hooks](https://docs.cypress.io/app/core-concepts/writing-and-organizing-tests#Hooks) of a test file.

File: pages/p/[id].tsx [hunk 6]
```diff
@@ -56,10 +56,9 @@ export const getServerSideProps: GetServerSideProps = async (context) => {
 const Post: React.FC<PropObj> = (props) => {
     let title = props?.post?.title
     let name = props?.post?.owner?.name
-
     if(props?.code) {
       return <Error statusCode={404} title="Post does not exist"/>
-    } else if(!props?.session?.user) {
+    } else if(!props?.session?.user || process.env.NEXT_PUBLIC_USE_TEST === 'false') {
```
Comment (alvinc21) [inline pages/p/[id].tsx]: I wonder if you could just pass a test session into the props so that you don't have to use the flag here.

File: lib/authSession.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { auth as useAuthSession } from '@/auth'
+
+export const auth = async (requestOrContext) => {
```
Comment (alvinc21) [inline lib/authSession.ts:3]: Is it possible to mock `requestOrContext` so that you can simplify this?

File: app/api/posts/route.ts [hunk 8]
```diff
@@ -5,7 +5,7 @@ import { prismaPost } from '@/lib/createPost'
 
 export const POST = auth(async (req) => {
     const data = await req.json()
-    if (req.auth) {
+    if (req.auth || process.env.NEXT_PUBLIC_USE_TEST === 'true') {
```
Comment (alvinc21) [inline app/api/posts/route.ts:8]: Is it possible to mock `req.auth` so you don't need to use the flag here?

File: cypress/support/commands.ts (hunk 3 above)
Comment (Chenxi96) [reply cypress/support/commands.ts]: No I don't think this is needed, will remove this file

File: cypress/support/e2e.ts (hunk 4 above)
Comment (Chenxi96) [reply cypress/support/e2e.ts]: No this isn't needed, Will remove this

File: components/CreateFormPost.tsx [hunk 9]
```diff
@@ -120,7 +120,7 @@ const CreateFormPost: React.FC<props> = (props) => {
         )
     } else {
         return (
-            <p>Post created</p>
+            <p className="PostCreated">Post created</p>
```
Comment (Chenxi96) [inline components/CreateFormPost.tsx:123]: added classname to specify the element for cypress testing.

File: cypress/e2e/HomePage.cy.ts [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    cy.visit('/')
+    cy.get('.createPost').click()
+    cy.get('#title').type('ball runs')
+    cy.get('#location').type('Air Canada Centre, 50 Bay St. #500, Toronto, ON M5J 2L2')
+    cy.get('#description').type('We need 4 more')
+    cy.get('#availableSpots').type('4')
+    cy.get('[type="datetime-local"]').type('2024-10-01T08:30')
+    cy.get('[type="submit"]').click()
+    cy.wait(500)
+    cy.get('.PostCreated').contains('Post created')
```
Comment (Chenxi96) [inline cypress/e2e/HomePage.cy.ts:29]: Specify the element with classname.

File: cypress.config.ts [hunk 11]
```diff
@@ -0,0 +1,40 @@
+import { defineConfig } from "cypress";
+import prisma from './lib/prisma'
+require('dotenv').config()
+
+export default defineConfig({
+  e2e: {
+    baseUrl: 'http://localhost:3000',
+    supportFile: false,
```
Comment (Chenxi96) [inline cypress.config.ts:8]: `supportFile` is set to false to remove the support folder with `e2e.ts` and `command.ts`

File: cypress.config.ts
Comment (Chenxi96) [inline cypress.config.ts:1]: added plugin event in `cy.task` to execute code in node for Prisma functions

File: app/api/posts/route.ts (hunk 8 above)
Comment (alvinc21) [reply app/api/posts/route.ts:8]: Put down what you found in your investigation here

File: lib/authSession.ts (hunk 7 above)
Comment (alvinc21) [reply lib/authSession.ts:3]: Put down what you found in your investigation here

**Review by alvinc21 — APPROVED** [review]

File: app/api/posts/route.ts (hunk 8 above)
Comment (Chenxi96) [reply app/api/posts/route.ts:8]: After attempting to mock `req.auth`, I realized that cypress only lets you mock the response object. As of right now, I am unable to mock the `req.auth` and the flag will be needed.

File: lib/authSession.ts (hunk 7 above)
Comment (Chenxi96) [reply lib/authSession.ts:3]: As of now it is a argument I am using, I don't have the knowledge to figure this out yet.
